tools/proofers/preview.js

```diff
@@ -64,6 +64,79 @@ function removeTrail(txt) {
     return txt;
 }
 
+const blockType = {
+    PARA: 0,
+    SECHEAD: 1,
+    SUBHEAD: 2,
+    HEAD: 3,
+    CONTINUATION: 4,
+};
+
+// split bText into blocks surrounded by blank lines
+// and process them by function procBlock
+function blockSplit(bText, procBlock) {
+    // find the number of blank lines preceding the block
+    // at beginning of page assume just past a single \n and thus in a paragraph
+    let prevEnd = -1;
+    // start is first non \n character
+    let beginRegex = /./g;
```

Comment:
in general, I put regular expressions outside functions to avoid them getting parsed and compiled every function invocation since that can be expensive. Obviously this doesn't apply if you are using some of the instance fields of a regular expression like lastIndex.

Reply:
Good plan. blockSplit only gets called once or twice per invocation of preview so it won't make much difference in this case.